docs: fix and clarify holistic example of handling storage (#2098)
canonical/operator: docs/howto/manage-storage.md

# How to manage storage
> See first: {external+juju:ref}`Juju | Storage <storage>`, {external+juju:ref}`Juju | Manage storage <manage-storage>`, {external+charmcraft:ref}`Charmcraft | Manage storage <manage-storage>`

## Manage storage for a machine charm

### Define the storage

> Define a `storage` section in `charmcraft.yaml` that lists the storage volumes and information about each storage. For example, for a transient filesystem storage mounted to `/cache/` that is at least 1GB in size:

Reviewer: I wonder if we should leave out the location entirely. It's not particularly useful to the charm, because the charm should always get the location from Juju, and it leads to bad behaviour where the charm hard-codes paths based on this instead of asking Juju.

```yaml
storage:
  cache:
    type: filesystem
    description: Somewhere to cache files locally.

>     location: /cache/

Author: It would be `/cache/0`, I think. Before my changes, there was a part that explained what the actual path would be, and why there would be numbers. I felt it was too distracting at this point, so removed it. (Why do I really care about the exact structure under`/cache/` if I'm not going to hardcode the path in my charm?)

But I'd be happy to clarify/reinstate the details

    minimum-size: 1G
    properties:
      - transient
```

> If you don't specify a location for the storage, Juju mounts the storage at the default location:

Author: Yeah, not obvious what the right thing is here... We haven't introduced the concept of additional storage instances yet. So perhaps the best wording at this point is:

> If you don't specify a location for the storage volume, ...

Matching the wording at the very beginning of the doc

```text
/var/lib/juju/storage/<storage-name>/0
```

### Access the storage

In your charm's `__init__` method, observe the [storage-attached](ops.StorageAttachedEvent) event:

```python
framework.observe(self.on["cache"].storage_attached, self._update_configuration)
```

In this example, we use a holistic event handler called `_update_configuration`. Alternatively, you could use a dedicated handler for the storage-attached event. To learn more about the different approaches, see [](#holistic-vs-delta-charms).

> Next, access the storage in `_update_configuration`. Don't hard-code the location of the storage. Instead, get the location from [`self.model.storages`](ops.Model.storages):

Reviewer: Similarly here, I think we should emphasise not hard-coding the location.

```suggestion
Next, access the storage in `_update_configuration`. **Don't hard-code the location of the storage.** Instead, get the location from [`self.model.storages`](ops.Model.storages):
```

```python
def _update_configuration(self, event: ops.EventBase):
    """Update the workload configuration."""
    cache = self.model.storages["cache"]
    if not cache:
        # The storage-attached event hasn't happened yet.
        logger.info("Storage is not yet ready.")
        return
    # Storage is available. Write some data to the cache.

>     cache_root = pathops.LocalPath(cache[0].location)

Reviewer 2: I find it a little weird that we check `if cache`, but then access `cache[0]`.

That's valid for unitary storage (actually very common in charms), but what about multiple storage units?

Reviewer: Perhaps instead of writing data to the cache storage, we could set it up in some way, and ensure we do that for all the "cache" storage attached. Something like:

```python
# Ensure that the storages are set up.
for storage in cache:
    cache_root = pathops.LocalPath(storage.location)
    (cache_root / "images").mkdir(exist_ok=True)
    (cache_root / "videos").mkdir(exist_ok=True)
```

>     (cache_root / "data.json").write_text("...")

Reviewer 2: Would be nice to show what "cache_root" may look like here.

```

> This example uses {external+charmlibs:ref}`charmlibs-pathops <charmlibs-pathops>` to write data to the storage.

Author: That would probably make sense. Storage (= "storage volume"?) is the abstract concept, "storage unit" is the directory on disk that you actually write data to

Instead of directly writing data to the storage, your charm could put the storage location in a configuration file. For example:

```python
    ...
    # Storage is available. Pass the storage location to the workload.
    config_path = pathops.LocalPath("/etc/my-app/cache-path.config")

>     config_path.write_text(cache[0].location)

Reviewer: And perhaps we could use all the storages here too, to make it clear that it's not always `[0]`. Something like:

```suggestion
    config_path = pathops.LocalPath("/etc/my-app/cache.config")
    config_path.write_text(f"paths = { ','.join(s.location for s in cache) }")
```

# How to manage storage
> See first: {external+juju:ref}`Juju | Storage <storage>`, {external+juju:ref}`Juju | Manage storage <manage-storage>`, {external+charmcraft:ref}`Charmcraft | Manage storage <manage-storage>`

## Manage storage for a machine charm

### Define the storage

> Each storage can be defined as supporting a single or multiple storage instances. If you define a storage as supporting multiple instances, your charm's users can use `juju add-storage` to increase the number of instances attached to your charm. (This command doesn't 'grow' existing instances).

Reviewer 2: ```suggestion
Each storage can be defined as supporting a single or multiple storage instances. If you define a storage as supporting multiple instances, your charm's users can use `juju add-storage` to increase the number of instances attached to the current unit. (Note that command doesn't 'grow' existing instances).
```

Juju also mounts the storage instance in the workload container's filesystem, at `/var/cache`. Depending on the workload, your charm might need to configure the workload to expect storage at this location.

### Configure the workload or access the storage

In your charm's `__init__` method, observe the [storage-attached](ops.StorageAttachedEvent) event:

```python
    framework.observe(self.on["cache"].storage_attached, self._update_configuration)
```

In this example, we use a holistic event handler called `_update_configuration`. Alternatively, you could use a dedicated handler for the storage-attached event. To learn more about the different approaches, see [](#holistic-vs-delta-charms).

Next, in `_update_configuration`, get the storage instance path in the workload container:

```python
def _update_configuration(self, event: ops.EventBase):
    """Update the workload configuration."""
    cache = self.model.storages["cache"]
    if not cache:
        logger.info("No instance available for storage 'cache'.")
        return
    web_cache_path = self.meta.containers["web"].mounts["cache"].location
    # Configure the workload to use the storage instance path.

>     ...

Author: In my simple test charm, storage was always available when my holistic handler looked for it earlier than storage-attached. However, I don't think this is guaranteed. I can't find any Juju docs that directly address this, [but there's a comment in the code](https://github.com/juju/juju/blob/3.6/internal/worker/uniter/storage/resolver.go):

> For CAAS models, we wait until after the start hook has run before running storage-attached hooks since storage is provisioned only after the pod has been created. [...] any charm logic that needs storage cannot be in install or start and if in config-changed, needs to be deferred until storage is available.

I've taken your suggestions and reworked the example like this:

```python
def _update_configuration(self, event: ops.EventBase):
    """Update the workload configuration."""
    cache = self.model.storages["cache"]
    if not cache:
        logger.info("No instance available for storage 'cache'.")
        return
    web_cache_path = self.meta.containers["web"].mounts["cache"].location
    # Configure the workload to use the storage instance path (assuming that
    # the workload container image isn't preconfigured to expect storage at
    # the location specified in charmcraft.yaml).
    # For example, provide the storage instance path in the Pebble layer.
    web_container = self.unit.get_container("web")
    try:
        web_container.add_layer(...)
    except ops.pebble.ConnectionError:
        logger.info("Workload container is not available.")
        return
    web_container.replan()
```

```

> See more: [](ops.Model.storages), [](ops.ContainerMeta.mounts)

> To access the storage instance in charm code, use one of the following approaches. The approaches are essentially equivalent - choose whichever you prefer.

Reviewer: They feel extra similar because `pathops` is hiding away that one is using Pebble (other than the error). I think if we're going to use `pathops` then we should just stick with one of them (and the local one probably makes more sense, because you don't need to worry about Pebble issues).

> To access the storage instance in charm code, use one of the following approaches. The approaches are essentially equivalent - choose whichever you prefer.

Author: On reflecting, this is less about the amount of data. The point is more that there might be other data in the workload container that you need to put into the cache (for example). In which case it makes more sense to do the file operations in the container. I've changed the wording to:

> This approach is more appropriate if you need to reference additional data in the workload container.

- Access the instance in the workload container, using {external+charmlibs:ref}`pathops <charmlibs-pathops>`. For example:

    ```python
        # Prepare the storage instance for use by the workload.
        web_container = self.unit.get_container("web")
        web_cache_root = pathops.ContainerPath(web_cache_path, container=web_container)
        try:
            (web_cache_root / "uploaded-data").mkdir(exist_ok=True)
            (web_cache_root / "processed-data").mkdir(exist_ok=True)
        except ops.pebble.ConnectionError:
            logger.info("Unable to access container. Waiting for pebble-ready.")
            return
    ```

- Access the instance in the charm container, using {external+charmlibs:ref}`pathops <charmlibs-pathops>` or standard file operations. For example:

    ```python
        # Prepare the storage instance for use by the workload.
        charm_cache_path = cache[0].location  # Always index 0 in a K8s charm.
        charm_cache_root = pathops.LocalPath(charm_cache_path)
        (charm_cache_root / "uploaded-data").mkdir(exist_ok=True)
        (charm_cache_root / "processed-data").mkdir(exist_ok=True)
    ```

## Handle storage detaching

> In the `src/charm.py` file, in the `__init__` function of your charm, set up an observer for the detaching event associated with your storage and pair that with an event handler. For example:

Reviewer 2: mention that it's only for machines... if that's only for machines

> observe the `storage-attached` event and handle any remaining setup once Juju

Reviewer: I can't attach this to the right line because it hasn't changed, but I believe there should be a "range" for the "scratch" storage, like the one above it. "multiple: range: 1-10" rather than "multiple: 1-10". I would probably again remove the "location" field as well.

Just above that, there's also a "storage-attached" that should probably be "`storage-attached`".

has attached the new storage.

## Manage storage for a Kubernetes charm

### Define the storage

Define a `storage` section in `charmcraft.yaml` that lists the storage volumes and information about each storage. Also specify where to mount the storage in the workload container.

For example, for a transient filesystem storage that is at least 1GB in size:

```yaml
storage:
  cache:
    type: filesystem
    description: Somewhere to cache files locally.
    minimum-size: 1G
    properties:
      - transient

containers:
  web-service:
    resource: app-image
    mounts:
      - storage: cache
        location: /var/cache
```

Juju mounts the storage at `/var/cache` in the workload container.

### Access the storage

In your charm's `__init__` method, observe the [storage-attached](ops.StorageAttachedEvent) event:

```python
framework.observe(self.on["cache"].storage_attached, self._update_configuration)
```

In this example, we use a holistic event handler called `_update_configuration`. Alternatively, you could use a dedicated handler for the storage-attached event. To learn more about the different approaches, see [](#holistic-vs-delta-charms).

> Next, access the storage in `_update_configuration`. Don't hard-code the location of the storage. Instead, get the location from [`ContainerMeta.mounts`](ops.ContainerMeta.mounts) for the workload container:

Reviewer: I think we should emphasise that the location shouldn't be hard-coded. I don't think it needs to be an admonition box, but perhaps:

```suggestion
Next, access the storage in `_update_configuration`. **Don't hard-code the location of the storage.** Instead, get the location from [`ContainerMeta.mounts`](ops.ContainerMeta.mounts) for the workload container:
```

    except ops.pebble.ConnectionError:
        # Pebble isn't ready yet. Return and wait for the pebble-ready event.
        logger.info("Pebble is not yet ready.")
        return
```

This example uses {external+charmlibs:ref}`charmlibs-pathops <charmlibs-pathops>` to write data to the storage in the workload container.

Instead of directly writing data to the storage, your charm could put the storage location in a configuration file. For example:

```python
    ...
    # Storage is available. Pass the storage location to the workload.
    cache_in_container = self.meta.containers["web-service"].mounts["cache"]
    config_path = pathops.ContainerPath(
        "/etc/my-app/cache-path.config",
        container=self.unit.get_container("web-service"),
    )
    try:
        config_path.write_text(cache_in_container.location)
    except ops.pebble.ConnectionError:
        ...
    # Ask the workload to reload its configuration.
```

## Handle storage detaching

In the `src/charm.py` file, in the `__init__` function of your charm, set up an observer for the detaching event associated with your storage and pair that with an event handler. For example:

```

> self.framework.observe(self.on.cache_storage_detaching, self._on_storage_detaching)

Reviewer: ```suggestion
self.framework.observe(self.on["cache"].storage_detaching, self._on_storage_detaching)
```

```

> See more: [](ops.StorageDetachingEvent)

Now, in the body of the charm definition, define the event handler, or adjust an existing holistic one. For example, to warn users that data won't be cached:

```
def _on_storage_detaching(self, event: ops.StorageDetachingEvent):
    """Handle the storage being detached."""
    self.unit.status = ops.ActiveStatus("Caching disabled; provide storage to boost performance)
```

> Examples: [MySQL handling cluster management](https://github.com/canonical/mysql-k8s-operator/blob/4c575b478b7ae2a28b09dde9cade2d3370dd4db6/src/charm.py#L823), [MongoDB updating the set before storage is removed](https://github.com/canonical/mongodb-operator/blob/b33d036173f47c68823e08a9f03189dc534d38dc/src/charm.py#L596)

> ## Write unit tests

Reviewer: I think we could have a better test example here, where the interaction with the storage is in the charm (maybe actually the content from the examples above) rather than faking it all in the context manager. But that could also easily be left for a future improvement.
